File: Arrays/Three Number Sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

vector<vector<int>> threeNumberSum(vector<int> array, int targetSum) {
  // Write your code here.

  vector<vector<int>> Sum;

  sort(array.begin(), array.end());

  int first = 0;

  int index = 0;

  int result = 0, pivot = 0;

  for(int last = 1; last < array.size(); last++)
  {
    pivot = (array[first] + array[last]);
    pivot = -pivot;
    //cout << "pivot: " << pivot << endl;
    if(array[first] <= pivot)
    {
      for(int i = 1; i < last; i++)
      {
        if(pivot == array[i] && pivot != array[first]&& pivot != array[last])
        {
          Sum.push_back(vector<int>());
          //cout << pivot << endl;
          Sum[index].push_back(array[first]);
          Sum[index].push_back(pivot);
          Sum[index].push_back(array[last]);
          
          index++;
        }
      }
      
      if(last == array.size()-1){
        first++;
        last = first+1;
      }
      
    }
    
  }
  
  return Sum;
}

File: Arrays/ThreeNumberSum.vb
Imports System.Collections.Generic
Improts System.Linq

Module Module1

    Function ThreeNumberSum(ByVal array As List(Of Integer), ByVal targetSum As Integer) 
    As List(Of List(Of Integer))

        Dim Sum As New List(Of List(Of Integer))()

        array.Sort()

        Dim first As Integer = 0
        Dim index As Integer = 0, result As Integer = 0, pivot As Integer = 0

        For last As Integer = 1 To array.Count -1
            pivot = (array(first)) + (array(last))
            pivot = -pivot

            If array(first) <= pivot Then
                For i As Integer = 1 To last - 1
                    If pivot = array(i) AndAlso pivot <> array(first) 
                                        AndAlso pivot <> array(last) 
                    Then 

                        Sum.Add(New List(Of Integer)())
                        Sum(index).Add(array(first))
                        Sum(index).Add(pivot)
                        Sum(index).Add(array(last))
                        index += 1
                    End If
                Next

                If last = array.Count -1 Then
                    first += 1
                    last = first + 1
                End If
            End If
        Next 
        
        Return Sum
    End Function

    Sub Main()

        Dim array As New List(Of Integer)() From {12, 3, 1, 2, -6, 5, -8, 6}

        Dim targetSum As Integer = 0

        Dim result As List(Of List(Of Integer)) = ThreeNumberSum(array, targetSum)

        For Each item As List(Of Integer) In result
            Console.WriteLine("[{0}]", StringJoin(",", item))
        Next 
    End Sub
    
End Module



